[그리디] 이창희 Spring MVC 1,2단계 미션 제출합니다

chxghee opened the pull request:
안녕하세요 다빈님~~~ 
처음 뵙겠습니다 ㅎㅎ

2주만에 스터디를 재개 했는데, 뭔가 기다려 지기도 하면서 시간이 이렇게 빨랐나 하는 생각도 드네요 

이번 미션도 잘 부탁드립니다!!

리뷰 요청을 드리기 앞서 제가 스프링을 경험해본 정도를 말씀 드리겠습니다

> 1. 강의를 보며 혼자 공부한 경험이 있어요
> 2. 게시판 같이 간단한 CRUD 사이드 프로젝트를 혼자 만들어 본 경험이 있어요

그래서 기본적으로 스프링이 제공하는 기능들을 얕게 경험해 봤어요

하지만 구현에만 집중해서 **스프링이 내부적으로 어떻게 동작하는지와** 어떤 코드가 잘 작성된 코드인지, 스프링을 이용한 프로젝트의 테스트 방법 등등은 잘 모릅니다 😅

그래서 이번 스프링 스터디에서는 제 코드의 부족한 점과 위에서 언급한 제 부족한 부분들을 고민해 볼수 있는 시간을 가져 볼수 있도록 
고민할 거리들을 제시해 주시면 감사할 것 같아요 👍


--- 

이번 미션은 이전 미션보다 비교적 간단했는데요,

1. 컨트롤러를 구현해서 타임리프같은 정적 리소스를 응답하는 것과,
1. 예약 정보를 응답하는 API 를 구현하는 것 이었어요

제가 구현한 방법을 간단하게 설명하자면,

1을 구현하기 위해 `spring-web`과 `thyemleaf` 의존성을 추가하고 `HomeController` 를 만들어 웹 페이지 이동에 대한 요청을 처리하도록 했고,

2를 구현하기 위해 예약 정보를 담는 `Reservation 엔티티`와 예약 관련 API 들을 처리하는 `ReservationComtroller`를 작성해서
`GET: /reservations`에 대한 요청을 `ReservationResponse DTO` 로 응답하도록 코드를 작성했어요


제 코드에서 미흡한 점과,  고민해 볼만한 포인트 들이 있다면 리뷰해 주시면 감사하겠습니다~~

chxghee committed:
Feat: 1단계 구현

의존성 추가 - web, thymeleaf
homeController 추가

chxghee committed:
Docs: 2단계 리드미 작성

chxghee committed:
Feat: 2단계 예약 조회 API 구현

70825 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
```
`@RestController`와 `@Controller`의 차이점은 무엇인가요??

chxghee replied:
`@Controller`는 컨트롤러 메서드가 View의 이름을 반환하면  Spring이 해당 View를 렌더링된 HTML로 변환한 후,  HTTP 응답 바디에 담아 클라이언트에 전송하고,

`@ResponseBody`는 컨트롤러 메서드에서 객체를 반환하면 해당 객체를 `HttpMessageConvertor`가  JSON 등으로 변환을 하고 이 변환된 JSON 데이터를 HTTP 응답 바디에 담아 클라이언트에 전송하게 됩니다.

`@RestController`는  `@Controller` + `@ResponseBody`를 합친 것으로 컨트롤러 클래스 내 모든 메서드가 객체를 반환하면 Http 바디에 데이터를 담아 응답하게 하는 어노테이션입니다

70825 commented on build.gradle:
```diff
@@ -14,6 +14,8 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation 'io.rest-assured:rest-assured:5.3.1'
 }
```
아래 내용 간단히 찾아보면 좋아보입니다~
- runtimeOnly
- compileOnly
- implementation
- testImplementation

chxghee replied:
`runtimeOnly` -  런타임 시에만 필요한 의존성  (DB 관련)

`compileOnly` - 컴파일 시에만 필요한 의존성  (Lombok 등)

`implementation` -  런타임  + 컴파일 시 필요한 의존성

`testImplementation` - 테스트 시에만 필요한 의존성 

라고 정리할 수 있을것 같아요

70825 replied:
굿굿 나중에 팀프로젝트하면서 의존성 추가할 때 어디에 써야할지도 판단하시면 좋습니다~

70825 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,23 @@
+
+    @GetMapping("/reservations")
+    public ResponseEntity<List<ReservationResponse>> getReservations() {
+        List<ReservationResponse> reservationResponseList = reservations.stream()
```
요거는 스타일 차이일 수도 있는데, 저는 `List<ReservationResponse>`를 보여주는 대신 `ReservationResponses` 같은거로 한번 더 감싸줘서 만드는 편이에요

[장점]
나중에 ReservationResponses 생성 로직이 복잡해져도 Controller는 그 코드를 몰라도 된다

[단점]
생각해보니까 단점이 있는지 모르겠음 🤣

chxghee replied:
우선 이 부분은 
미션에서 제시된 API 명세가 리스트를 반환하는 형태라 일단은 List를 반환하도록 했어요

이것과 별개로 List를 한번 감싸서 ReservationResponses 응답하는 것은 여러 장점이 있는것 같네요!

1. 위에서 언급하셨듯 나중에 `ReservationResponses` 생성 로직이 복잡해져도 Controller는 그 코드를 몰라도 된다
2. `List<Reservation>`과 관련된 부가적인 정보(개수, 페이징 정보 등등)를 `ReservationResponses`에 담을 수 있어 확장성 있는 설계가 될수 있다

즉 정리하자면 요구사항이 달라져, API 스펙의 수정이 필요할 때 응답하는 데이터의 구조 자체가 변화하지 않아 클라이언트 쪽에서도 변경점이 적고, 훨씬 확장성 있는 설계가 되는 것 같아요

70825 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,23 @@
+
+    private List<Reservation> reservations = new ArrayList<>();
+
+    @GetMapping("/reservations")
```
이건 저도 최근에 알게 된건데 메서드 위에 적는가 / 클래스 위에 적는가도 스타일의 차이더라구요

저는 메서드 위에 두어야 검색하기 편하다고 생각했는데, `IntelliJ`에 EndPoints라는 기능 사용해서 URI 검색해서 클릭하면 관련 코드로 바로 접근이 가능하더라구요

클래스 위에 `@RequestMapping("/tmp")`를 추가한 상황에서 실행했습니다.

![스크린샷 2025-04-30 오후 9 56 22](https://github.com/user-attachments/assets/18471e5f-ac9f-4b59-8b28-1f214ad93ee1)

IntelliJ 오른쪽 맨 위 돋보기 > Endpoints > 위 사진처럼 보임

이미 알고 있던거라면 저한테 인텔리제이 유용한 팁 좀 알려주시죠 ㅋㅋㅋ

chxghee replied:
이 부분은 몰랐던 부분이네요

그렇다면 더욱 `@RequsetMapping`을 쓰지 않을 이유가 없겠네요 ㅎㅎ

Reservation Controller 클래스에  `@RequestMapping("/reservations")` 를 붙이는 걸로 수정해 볼게요 👍

인텔리제이는 저도 잘 몰라서  차차 알아가는 단계입니다..ㅋㅋㅋㅋㅌㅋ ㅠ
(사실 엔드포인트 검색하는 법도 처음 알았음)

70825 replied:
ㅋㅋㅋ 저는 참고로 저거 알고 있어도 그냥 메서드 위에 URI를 붙이고 cmd + Shift + F (= 윈도우 Ctrl + Shift + F)로 경로 전부 입력하고 접근하는게 이미 너무 익숙해져서 Endpoints는 HTTP 응답 테스트로 확인할 때만 사용하고 있습니다

이거는 창희님 하고 싶은 대로 수정하시져

70825 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,23 @@
+        List<ReservationResponse> reservationResponseList = reservations.stream()
+                .map(ReservationResponse::new)
+                .toList();
+        return ResponseEntity.ok(reservationResponseList);
```
현재 미션 단계에서는 중간에 에러가 나오든 뭐든 무조건 OK 응답만 보내는 것으로 보이는데요
만약 에러 메시지를 표기하게 된다면 어떤 내용을 구성해서 보여주는게 좋을까요?? 일단 기본적으로 상태코드는 보여준다고 가정할게요
코드 추가는 4단계 내용이라 어떤 내용으로 구성하는게 좋을까에 대해서만 이야기 해봅시답

chxghee replied:
위의 예약 정보들을 조회하는 API의 경우 `204 No Content` 를 응답하고,
메세지는 "현재 예약된 방탈출이 없습니다" 정도로 구성하면 될까요? 

 `404 Not Found` 도 선택지중 하나가 될 수 있을것 같네요! 

근데 4xx 에러는 클라이언트쪽의 잘못된 요청으로 발생한 에러라는 뉘양스가 나서 `204 NoContent`가 더 나은것 같다라는 생각이 드네요

70825 replied:
앗 추가 설명을 하는게 좋겠네요. `/reservations` 말고도 다른 다양한 URI가 있다고 가정해보면 에러를 보여줄 때 어떤 내용들을 구성해서 보여줄 수 있을까요?

지금 창희님이 이야기한 내용으로는 아래처럼 구성할 수 있겠네요
- 상태 코드
- 에러가 발생한 원인

이거 포함해서 더 추가적으로 넣을만한 정보가 있을까요? 아니면 이거로도 충분할까요?

---

> 위의 예약 정보들을 조회하는 API의 경우 204 No Content 를 응답하고,
메세지는 "현재 예약된 방탈출이 없습니다" 정도로 구성하면 될까요?
404 Not Found 도 선택지중 하나가 될 수 있을것 같네요!
근데 4xx 에러는 클라이언트쪽의 잘못된 요청으로 발생한 에러라는 뉘양스가 나서 204 NoContent가 더 나은것 같다라는 생각이 드네요

이 내용은 댓글을 보니 좋은 학습이 될 것 같아서 질문 드려요~

학습 1
1) `localhost:8080/reservation`을 입력하면 `/reservations`도 호출하는 구조인데요. 크롬 > 개발자 도구 > 네트워크를 사용해서 `/reservations`의 응답 패킷 body 값을 확인해봅시다
2) 현재 코드에 데이터가 없을 경우 no content를 응답하도록 코드를 수정해봅시다 (커밋 없이 코드 수정만)
3) 다시 1번 과정을 수행해서 `/reservations`의 응답 패킷 body 값을 확인해봅시다.
4) 응답 body 값에 차이가 있을까요? 다르다면 왜 다를까요?

학습 2
1) 이번엔 `localhost:8080/reservations`으로 직접적으로 접속해봅시다
2) no content로 수정할 때와 수정하기 전이랑 접속 했을 때 둘이 무슨 차이가 있을까요? 그리고 왜 이런 차이가 발생할까요?

chxghee replied:
다양한 URI 가 있다면 에러 응답 시 body에 추가적으로 넣을만한 정보에는

어떤 경로의 요청에서 응답된 에러인지 명확하게 보여 주기 위해  `요청이 들어온 경로` 를 같이 담아 보내주면 좋을것 같아요

또한 사용자에게 입력을 받아 요청을 처리하는 API 의 경우 사용자의 `잘못된 입력값`을 응답해 주면 좋을것 같아요~~


지금은 이렇게만 생각이 나는데, 혹시 이 외에도 추가적으로 넣을 만한 정보가 있다면 조언 부탁드려요 !!

---

<br>

## 학습 1

### 200 OK 응답
<img width="419" alt="image" src="https://github.com/user-attachments/assets/d73c93fe-7dcd-4049-92f8-d650fed2a7b2" />

<img width="346" alt="image" src="https://github.com/user-attachments/assets/148b52a9-70a1-40cb-a2bb-56cfc6322346" />

200 응답의 경우 예약이 없을때는 빈 배열이 body에 담겨 응답되지만,

### 204 No Content 응답
<img width="445" alt="image" src="https://github.com/user-attachments/assets/3c6f4a2e-0710-464a-9f6a-e0c4c650b6ac" />

<img width="594" alt="image" src="https://github.com/user-attachments/assets/c8a07315-e80c-4317-9067-6946f2187bad" />

`204 No Content`는 코드 이름과 알맞게 body 자체가 존재하지 않는군요!

이렇게 되면, `204 No Content`는 에러시 에러가 발생한 원인 등 다른 메세지를 바디에 넣어 보내줄 수 없게 되겠네요


## 학습 2

### 200 OK 응답

<img width="394" alt="image" src="https://github.com/user-attachments/assets/eb67e4ad-582b-4d63-b325-6abe687e94d6" />

200 OK 는 응답 body의  빈 배열을 렌더링해서, 화면에 보여 주는데,

### 204 No Content 응답

`204 No Content`는 바디 자체가 없기 때문에 어떠한 view도 그려줄 수 없어서 이전 화면이 그대로 보이네요

그렇다면 명확하게 오류의 발생 원인을 사용자에게 알리려면 `404 Not Found` + Body 에 관련한 메세지를 담아 응답하는게 좋갰네요!

Q. 그렇다면 반대로 `204 No Content` 는 응답 body에 메세지를 담을 수 없다는 큰 단점 때문에 실무에 거의 사용하지 않나요??

70825 replied:
> 다양한 URI 가 있다면 에러 응답 시 body에 추가적으로 넣을만한 정보에는
어떤 경로의 요청에서 응답된 에러인지 명확하게 보여 주기 위해 요청이 들어온 경로 를 같이 담아 보내주면 좋을것 같아요
또한 사용자에게 입력을 받아 요청을 처리하는 API 의 경우 사용자의 잘못된 입력값을 응답해 주면 좋을것 같아요~~
지금은 이렇게만 생각이 나는데, 혹시 이 외에도 추가적으로 넣을 만한 정보가 있다면 조언 부탁드려요 !!

```
- 상태코드
- 에러가 발생한 원인
- 잘못된 입력값
```

오호 좋습니다. 저는 지금 요구사항에서 필요한 내용은 다 보여진 것 같아요
아래 내용은 저도 이번주에 학습한 내용이긴한데, 에러 메시지도 표준이 있더라구요
그래서 아래 내용 심심하실때 이런 정보들도 보여주는구나하면서 읽어보면 좋아보여요

- [RFC 7807](https://datatracker.ietf.org/doc/html/rfc7807) > Section 4까지 읽으면 됩니다
- [Spring Web Servlet - DefaultErrorAttributes](https://docs.spring.io/spring-boot/api/java/org/springframework/boot/web/servlet/error/DefaultErrorAttributes.html) > RFC 9457 (= RFC 7807의 최신화)를 지원하는 에러 핸들링

---

> 학습 1
204 No Content는 코드 이름과 알맞게 body 자체가 존재하지 않는군요!
이렇게 되면, 204 No Content는 에러시 에러가 발생한 원인 등 다른 메세지를 바디에 넣어 보내줄 수 없게 되겠네요
학습 2
204 No Content 응답
204 No Content는 바디 자체가 없기 때문에 어떠한 view도 그려줄 수 없어서 이전 화면이 그대로 보이네요
그렇다면 명확하게 오류의 발생 원인을 사용자에게 알리려면 404 Not Found + Body 에 관련한 메세지를 담아 응답하는게 좋갰네요!

네 맞아요 No Content 사용처는 정해져 있기 때문에 [RFC 7231 - 204 No Content](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.5) 읽어보면 좋아보입니답
추가로 200 OK도 읽으면 좋아보여요

그리고 404 Not Found + Body는 창희님이 코멘트한 `근데 4xx 에러는 클라이언트쪽의 잘못된 요청으로 발생한 에러라는 뉘양스가 나서 204 NoContent가 더 나은것 같다라는 생각이 드네요` 이 내용이 맞기 때문에 저는 200번대로 상태코드 응답하는게 맞다고 생각해요. 클라이언트에서는 정상요청을 했지만, 서버에서 조회해보니 데이터가 없는데, 데이터가 없는 곳에 요청을 보낸 클라이언트 잘못이라고 하는건 좀 이상한거니까요 

>  Q. 그렇다면 반대로 204 No Content 는 응답 body에 메세지를 담을 수 없다는 큰 단점 때문에 실무에 거의 사용하지 않나요??

아뇨 사용 용도가 정해져 있기 때문에 사용하고 있어요. 요것도 위 문서 읽고 예시들 찾아보시면 같이 해결될 것으로 보여요

70825 commented on src/main/java/roomescape/entity/Reservation.java:
```diff
@@ -0,0 +1,36 @@
+
+    private Long id;
+    private String name;
+    private LocalDate date;
+    private LocalTime time;
```
이거는 `LocalDateTime`으로 하나로 합칠 수 있을 것 같은데, `date`, `time`이 나눠진 이유가 있을까요?

chxghee replied:
우선 미션의 데이터 입력 형태가 날짜를 입력하고 시간을 입력하는 형태라 날짜와 시간이 따로 처리 되고 있고, 
조회 시에도 따로 따로 표현되고 있어요

이러한 상황이라면 LocalDateTime으로 하나로 합치기 보다는 따로 관리하는 것이 데이터를 따로 포맷팅하지 않아도 되고 편리할 것 같아서 나누었어요

그런데 지금 든 생각이 예약의 시간 관련 정보를 처리 할 요구사항이 있다면,
시간과 날짜 값이 따로 저장되기 때문에 날짜와 시간 정보는 혼자서는 의미가 없어 지고 항상 서로가 같이 조회 되어야 의미가 있는 상황이 고 변경이 되거나 한다면 시간 정보의 일관성이 위험해 질 수도 있을 것 같다라는 생각이 들었어요

그렇다면 데이터는 LocalDateTime 으로 저장하고 입출력은 LocalDate와 LocalTime으로 분리하는 것도 하나의 방법이 될 수 있겠군요

장단점이 있는것 같은데, 이에 대해서는 어떻게 생각하시나요??

70825 replied:
> 이러한 상황이라면 LocalDateTime으로 하나로 합치기 보다는 따로 관리하는 것이 데이터를 따로 포맷팅하지 않아도 되고 편리할 것 같아서 나누었어요

좋습니다~

> 그런데 지금 든 생각이 예약의 시간 관련 정보를 처리 할 요구사항이 있다면,
시간과 날짜 값이 따로 저장되기 때문에 날짜와 시간 정보는 혼자서는 의미가 없어 지고 항상 서로가 같이 조회 되어야 의미가 있는 상황이 고 변경이 되거나 한다면 시간 정보의 일관성이 위험해 질 수도 있을 것 같다라는 생각이 들었어요

`일관성이 위험해질 수도 있다 =  개발자가 깜빡하고 시간, 날짜 둘중에 하나를 업데이트를 하지 않는다`로 이해해도 될까요?

> 장단점이 있는것 같은데, 이에 대해서는 어떻게 생각하시나요??

음 이번 미션 10단계까지 살펴봤는데, 요구사항대로 하려면 `LocalDate`, `LocalTime` 두 개로 나눌 수 밖에 없겠군요
저는 `LocalDateTime`이 필요해질 부분을 기대 했는데, 없어 보여서 지금은 `LocalDateTime`은 단점 밖에 없을 것 같아요

chxghee replied:
> `일관성이 위험해질 수도 있다 =  개발자가 깜빡하고 시간, 날짜 둘중에 하나를 업데이트를 하지 않는다`로 이해해도 될까요?

-> 네 이런 생각이 들었었어요 말을 좀 애매하게 했었네요 😅

> 저는 LocalDateTime이 필요해질 부분을 기대 했는데, 없어 보여서 지금은 LocalDateTime은 단점 밖에 없을 것 같아요

그렇다면 혹시 어떤 상황에서는 `LocalDate`,  `LocalTime`으로 따로 관리하는게 좋고, 아니면 `LocalDateTime` 처럼 합쳐서 관리하는게 좋은지 알려주실수 있나요? 다빈님의 생각이 궁금합니다!!

70825 replied:
Spring JDBC 5단계를 보면 디비 테이블이 date, time으로 나눠져 있더라구요
근데 저는 `LocalDate`, `LocalTime` 이걸 도메인, 엔티티, 디비까지 따로 관리하는걸 이번에 처음 봤어요 ㅋㅋ
이 미션 자체를 우테코에서 진행하지 않았습니다 (= 기수마다 미션이 조금씩 다름)
그래서 `어떤 상황에서는 LocalDate, LocalTime으로 따로 관리하는게 좋고` 여기에 대한 상황 자체를 경험해보지 못했습니답

`LocalDateTime 처럼 합쳐서 관리하는게 좋은지`에 대한 내용은 보통 시간 데이터는 날짜가 없으면 의미 없는 상황이 대부분이기도 하고, `LocalDateTime` 하나로 저장할 수 있는걸 `LocalDate`, `LocalTime`으로 나눠서 저장할 이유가 보이지 않아보여요

`LocalDate`, `LocalTime`과 관련되서 다음 3, 4단계 미션에서 살짝 주의할 점이 하나 있어보이긴하는데, 이건 창희님의 코드에 따라 달라지는거라 스포하면 안되니 다음에 이야기할게요 ~

70825 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -5,6 +5,8 @@
+import static org.hamcrest.Matchers.is;
+
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
`@DirtiesContext` 코드를 뜯어보면 이런 코드가 있는데요

```java
    public static enum ClassMode {
        BEFORE_CLASS,
        BEFORE_EACH_TEST_METHOD,
        AFTER_EACH_TEST_METHOD,
        AFTER_CLASS;

        private ClassMode() {
        }
    }
```

`@DirtiesContext`가 무슨 역할을 하는지, 각 ClassMode는 어떤 기능을 하는지 찾아보면 좋아보입니다

나중에 테스트 코드를 많이 작성하게 된다면 `@DirtiesContext`는 체감이 많이 날 정도로 성능이 느려서 사용하지 않긴 하는데요. 근데 나중에 테스트 격리라는 개념을 공부하게 된다면 어떤 과정으로 격리를 해볼 수 있을까에 대한 아이디어는 여기서 따올 수 있어보입니다.

나중에 팀프로젝트에서 테스트 코드 많이 만들게 될 때, 지금 리뷰가 기억이 난다면 https://youtu.be/N06UeRrWFdY?si=Ks70VFuOQd4EhM9P 이 영상도 보는걸 추천드려요

chxghee replied:
이번에 처음 접해 봐서 조금 찾아 봤습니다!

`@DirtiesContext`는 어플리케이션 컨텍스트를 공유하게 되는 테스트들을 격리하기 위한 어노테이션입니다.

이 어노테이션을 사용하게 되면, 테스트를 실행하기 전 후에 컨텍스트를 초기화 할 수 있어요

하지만 매번 컨텍스트를 초기화 하기 때문에 테스트의 성능에 영향을 끼칠수 있다고 하네요

ClassMode는 어떤 시점에 컨텍스트를 초기화 할지 지정하는 옵션입니다

`BEFORE_CLASS`: 테스트 클래스의 모든 메서드 실행 전에 컨텍스트를 초기화
`AFTER_CLASS`: 테스트 클래스의 모든 메서드 실행 후에 컨텍스트를 초기화
`BEFORE_EACH_TEST_METHOD`: 각 테스트 메서드 실행 전에 컨텍스트를 초기화
`AFTER_EACH_TEST_METHOD`: 각 테스트 메서드 실행 후에 컨텍스트를 초기화

위와같이 네 가지 옵션이 있어요

00_CLASS는 모든 테스트를 클래스 단위로 실행할때만 적용이 되고 개별 메서드 별로 실행할 때는 적용이 안됩니다

혹시 놓친 중요한 부분이 있다면 피드백 부탁드려요!!

70825 replied:
> 하지만 매번 컨텍스트를 초기화 하기 때문에 테스트의 성능에 영향을 끼칠수 있다고 하네요

여기서 컨텍스트 초기화가 무슨 뜻이라고 생각하시나요??

chxghee replied:
우선 컨텍스트 초기화에 대해 말하기 전에 어플리케이션 컨텍스트부터 개념을 정리해 볼게요


### Application Context

> 어플리케이션 컨텍스트는 Spring 어플리케이션이 돌아가는 "상태가 있는" 실행 환경 전체를 의미

컨텍스트가 맡은 주요 업무는 

1. `Bean` 을 생성, 주입, 관리를 합니다 (빈의 라이프 사이클을 관리한다 라고 할 수 있겠네요)

2. `yml` 파일이나 `@Configuration` 클래스들의 설정들을 읽어 실행되는 어플리케이션에 적용합니다

3. 이벤트 처리
ApplicationEventPublisher를 통해 이벤트를 발행하고, 리스너가 이를 수신


이렇듯 컨텍스트는 어플리케이션의 전반적인 상태를 관리하게 됩니다

그렇기 떄문에 컨텍스트의 생성은 매우 볼륨이 크고 무거운 작업일 수 밖에 없습니다
이를 개선하기 위해 Spring에선 내부적으로 컨텍스트를 캐싱해서 사용하는 선택을 하였는데,


이러한 캐싱으로 인한 컨텍스트의 상태 유지가 테스트 별로 독립적인 테스트를 힘들게 하는 상황이 발생할 수 있습니다
이를 해결하기 위해 `@DirtiesContext`을 사용해 테스트의 실행 이후 혹은 이전에 컨텍스트를 초기화 하여 테스트의 독립성을 보장할 수 있어요 
(하지만 캐시를 비우고 컨텍스트를 다시 만들기 때문에  성능에 영향을 끼칠수 있어요)

즉 `컨텍스트의 초기화`란 위에서 말한 어플리케이션의 상태가 있는 실행 환경 전체를 초기 상태로 되돌리는 것 

다시말해 _"컨텍스트를 재생성해서 어플리케이션을 처음부터 다시 시작하는 것"_ 이라고 할 수 있을 것 같아요

부족한 내용이 있다면 피드백 부탁드려요!

70825 replied:
굿굿 좋습니다
ApplicationContext가 저런 일을 하기 때문에 생각보다 시간이 오래 걸립니다
테스트가 많아지게 되고, 한 개의 테스트가 끝날 떄마다 컨텍스트 초기화를 진행하면 테스트 전체 실행하는데 분 단위가 걸리는 매직을 볼 수 있어요. 제가 팀프로젝트할 때는 `@DirtiesContext`로 변경해서 전체 테스트 실행해봤더니 완료하는데 5분 이상이 걸리더라구요 (캐싱 적용시 30초 미만)

참고로 나중에 스프링 관련 서적 읽을 때, 옛날 서적에서는 BeanFactory라는 단어도 나올텐데요. 이건 ApplicationContext의 이전 버전이라고 생각하시면 됩니답

70825 commented on src/main/java/roomescape/controller/HomeController.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
+public class HomeController {
```
`HomeController`와 `ReservationController`를 보면서 둘이 구분을 해야한다는 생각이 들었는데요

[생각이 든 상황]
만약 이런 어드민 페이지를 보여주는 컨트롤러가 10개가 넘어가고, `ReservationController`처럼 API를 제공하는 컨트롤러도 10개가 넘어간다고 해볼게요.
그러면 controller 파일 안에 두가지 종류의 컨트롤러가 아무런 분류 없이 계속 추가되고 있는데요

어떻게 하면 사용자가 두가지 종류의 컨트롤러를 나눠서 저장해 볼 수 있을까요?

[이해를 쉽게하기 위한 예시]
네이밍 아무거나 막 붙였습니다
![스크린샷 2025-04-30 오후 10 25 48](https://github.com/user-attachments/assets/c6a68b39-6c39-4113-9013-d19ae8cf2788)

chxghee replied:
컨트롤러에 패키지를 view / api 로 분리해서 저장하면 훨씬 나을것 같아요

근데 view라는 이름은 좀 어색하게 느껴지는것 같기도 한데 어떻게 생각하시나요..? ㅋㅋㅋㅋ 😅

70825 replied:
저는 회사에서 view가 prefix인 패키지 네이밍이 존재해서 괜찮아보여요
1) controller > view / api로 나눠서 관리
2) viewcontroller / apicontroller 나눠서 관리

로 수정해볼 방법이 두가지 방법이 보입니답

70825 reviewed (changes requested):
안녕하세요 창희님~ 저도 잘 부탁드립니다 ㅋㅅㅋ

> 리뷰 요청을 드리기 앞서 제가 스프링을 경험해본 정도를 말씀 드리겠습니다
1.강의를 보며 혼자 공부한 경험이 있어요
2.게시판 같이 간단한 CRUD 사이드 프로젝트를 혼자 만들어 본 경험이 있어요
그래서 기본적으로 스프링이 제공하는 기능들을 얕게 경험해 봤어요
하지만 구현에만 집중해서 스프링이 내부적으로 어떻게 동작하는지와 어떤 코드가 잘 작성된 코드인지, 스프링을 이용한 프로젝트의 테스트 방법 등등은 잘 모릅니다 😅

제가 스프링 공부할 때 아래 여러가지 책을 읽었는데요
1) 개발자가 반드시 정복해야할 객체지향과 디자인 패턴
2) 스프링 입문을 위한 자바 객체 지향의 원리와 이해
3) 초보 웹 개발자를 위한 스프링 5 프로그래밍 입문

이중에서 책 읽을 시간이 되신다면 3, 2를 추천드리고, 이중에서도 3번을 제일 추천합니다
우테코 5기 할 때는 3번 책 읽는 크루들이 많았었고, 저도 그중 한명이었는데, 우테코 하는 동안 읽은 서적중에 제일 유익 했었어요 (1번 책은 4기 추천 도서로 기억합니다 - 3번 책이랑 같은 저자에요)
참고로 1,3 책은 저자분이 유튜브랑 블로그도 하는데 도움 되는 영상이 많으니 심심할 때 참고바랍니다
https://www.youtube.com/@madvirus/videos

> 제 코드에서 미흡한 점과, 고민해 볼만한 포인트들이 있다면 리뷰해 주시면 감사하겠습니다~~

일단 미션이 많이 간단하기 때문에 코드 리뷰랑 테스트코드에 대한 리뷰가 좀 힘드네요 ㅋㅋ…

그래서 대신 고민해볼만한 포인트 위주로 최대한 많이 남겼습니다

코드 변경할 부분도 당장은 고칠 부분이 없고, 이야기하다가 생길 수도 있는 부분이라 일단은 코멘트 남기면서 같이 생각해보시죠!

chxghee committed:
Refactor: 리뷰 반영
ReservationApiController URI mapping  @RequsetMapping 으로 변경
컨트롤러 패키지 세분화 view, api

70825 reviewed (approved):
고생하셨습니다~ 🚀🚀
